rfd/9999-aws-api-integration-using-oidc.md

```diff
@@ -0,0 +1,621 @@
+The only updatable field is `aws_role_arn`.
+
+HTTP API:
+```
+Methods:
+GET .../integration
+GET .../integration/:id
+POST .../integration
+PUT .../integration/:id
+DELETE .../integration/:id
+```
+JSON representation:
+```json
+{
+	"name": "myaws",
+	"subkind": "aws-oidc"
+	"subkindData": {
+		"aws_role_arn": "arn:aws:123:TeleportOIDC"
+	}
+}
+```
+
+In order to call an Integration action, we'll create a new Web API endpoint:
+```
+--->
+POST .../integration/:id/action/aws-oidc-list-databases
+{}
+
+<---
+{
+    "status": "success",
+    "response": {
+        "items": [
+            {
+                "status": "available",
+                "name": "marcodbtest",
+                "iamAuth": "true",
+                "engine": "postgres",
+                "engineVersion": "15.2",
+                "masterUsername": "postgres",
+                "tags": [],
+                "arn": "arn:aws:rds:us-east-1:<accid>:db:marcodbtest",
+                "addr": "marcodbtest.<someid>.us-east-1.rds.amazonaws.com",
+                "port": "5432"
+            }
+        ]
+    }
+}
+```
+
+This will generate a new token for the Integration `:id` and then call the `aws-oidc-list-databases` method which maps to a method that will do two AWS API Calls:
+- `rds describe-db-instances`
+- `rds describe-db-clusters`
+
+The response contains the following fields:
+- `status`: reports the result of calling the integration (other errors will come as usual: http status code)
+- `response`: specific to the requested action
+
+Pagination/Offset and search capabilities are out of scope for now.
+
+#### CLI
+Users can also create/update the resource using `tctl`.
+
+The only updatable field is `aws_role_arn`.
+
+`tctl`:
+
+```
+$ tctl get integrations
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+
+$ tctl get integration/myaws
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: some-name
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+
+$ tctl create aws-integration.yaml
+```
+
+Resource representation as `yaml`:
+```yaml
+kind: integration
+subkind: aws-oidc
+version: v1
+metadata:
+	name: myaws
+spec:
+	aws_role_arn: arn:aws:123:TeleportOIDC
+```
+
+#### IaC - Terraform
+A new resource `Integration` must be created to allow for the Integration management from the Terraform provider.
+
+#### IaC - Kube Operator
+A new resource `Integration` must be created to allow for the Integration management from the Kube Operator.
+
+#### IaC - Helm Charts
+No change will be made to the Helm Charts because there's no new configuration changes.
+
+### Security
+
+#### Rotation
+JWT Signing keys can be rotated using `tctl auth rotate --type=oidc`.
+
+JWKS endpoint returns a list of public keys, so the old and the new key are provided for a seamless migration.
+
+#### MITM between AWS and Teleport
+
+The only configuration AWS has about the provider is its URL.
```

Comment:
As noted in my lower comment, this isn't strictly true, AWS also fingerprints the top intermediate certificate authority that issued the IdP's TLS certificate.